client/core/core.go

```diff
@@ -0,0 +1,251 @@
+// This code is available on the terms of the project LICENSE.md file,
+// also available online at https://blueoakcouncil.org/license/1.0.0.
+
+package core
+
+import (
+	"context"
+	"encoding/json"
+	"net/url"
+	"sync"
+	"time"
+
+	"decred.org/dcrdex/client/comms"
+	"decred.org/dcrdex/client/db"
+	"decred.org/dcrdex/client/db/bolt"
+	"decred.org/dcrdex/dex"
+	"decred.org/dcrdex/dex/msgjson"
+)
+
+var log dex.Logger
+
+// websocket is satisfied by a comms.WsConn, or a stub for testing.
+type websocket interface {
+	NextID() uint64
+	WaitForShutdown()
+	Send(msg *msgjson.Message) error
+	Request(msg *msgjson.Message, f func(*msgjson.Message)) error
+	MessageSource() <-chan *msgjson.Message
+}
+
+// dexConnection is the websocket connection and the DEX configuration.
+type dexConnection struct {
+	conn   websocket
+	assets map[uint32]*msgjson.Asset
+	cfg    *msgjson.ConfigResult
+}
+
+// Config is the configuration for the Core.
+type Config struct {
+	// DBPath is a filepath to use for the client database. If the database does
+	// not already exist, it will be created.
+	DBPath string
+	// Logger is a logger for the core to use. Having the logger as an argument
+	// enables creating custom loggers for use in a GUI interface.
+	Logger dex.Logger
+	// Certs is a mapping of URL to filepaths of TLS Certificates for the server.
+	// This is intended for accomodating self-signed certificates, mostly for
+	// testing.
+	Certs map[string]string
+}
+
+// Core is the core client application.
+type Core struct {
+	ctx     context.Context
+	wg      sync.WaitGroup
+	cfg     *Config
+	connMtx sync.RWMutex
+	conns   map[string]*dexConnection
+	db      db.DB
+	certs   map[string]string
+}
+
+// New is the constructor for a new Core.
+func New(cfg *Config) *Core {
+	log = cfg.Logger
+	core := &Core{
+		cfg:   cfg,
+		conns: make(map[string]*dexConnection),
+	}
+	return core
+}
+
+// Run runs the core. Satisfies the runner.Runner interface.
+func (c *Core) Run(ctx context.Context) {
+	log.Infof("started DEX client core")
+	db, err := bolt.NewDB(ctx, c.cfg.DBPath)
+	if err != nil {
+		log.Errorf("database initialization error: %v", err)
+		return
+	}
+	c.ctx = ctx
+	c.db = db
+	go c.initialize()
+	c.wg.Wait()
+	log.Infof("DEX client core off")
+}
+
+// ListMarkets returns a list of known markets.
+func (c *Core) ListMarkets() []*MarketInfo {
+	c.connMtx.RLock()
+	defer c.connMtx.RUnlock()
+	var infos []*MarketInfo
```

Comment:
Can set size here?
`infos := make([]*MarketInfo, 0, len(c.conns))`
For some reason I think this is good practice, but I`m not sure. Sorry, more of a question than a suggestion.